### PR TITLE
Remove some `gmp.h` inclusions and cleanup some headers

### PR DESCRIPTION
**Pull request by albinahlback**

### CODE REVIEW

**Comment by fredrik-johansson** [comment]
Not sure if ``fmpz_set_uiui`` and ``fmpz_get_uiui`` and friends should be completely deinlined. They feature in some rather important loops.

**Comment by albinahlback** [comment]
> Not sure if `fmpz_set_uiui` and `fmpz_get_uiui` and friends should be completely deinlined. They feature in some rather important loops.

Sure. I'll try to figure it out.

**Comment by albinahlback** [comment]
Does this work? I assume that the fmpz-collector never pushes anything with less than two limbs. I will just check that it indeed is the case for all files in `fmpz/link/`.

**Comment by albinahlback** [comment]
The PR got a little bit bigger than I intended. Notably, I inserted an assertion that any returned mpz in `_fmpz_clear_mpz` has at least two allocated limbs.

Just as to note: Currently, with `FLINT_TEST_MULTIPLIER` larger than something around 5.5, the test for `fq_nmod_mpoly_factor` fails due to a multivariate polynomial not being canonical. I don't know if that is present in the current state of the master branch, but this is something to look into.

**Comment by albinahlback** [comment]
Perhaps we should increase the test multipliers for runners with assert.

File: doc/source/fmpz.rst [hunk 1]
```diff
@@ -130,16 +130,35 @@ Types, macros and constants
...
 
+.. note::
+
+   As of FLINT 3.2.0, it is required that the *mpz* pointed to by *f* has at
+   least :macro:`MPZ_MIN_ALLOC` limbs allocated when :func:`_fmpz_clear_mpz` is
+   called. Note that GMP functions never reduces the number of allocated limbs,
+   apart from a bug in ``mpz_remove`` which as of GMP 6.3.0 may reduce the
+   number of allocated limbs in the output *mpz*. Hence, as long as
+   ``mpz_realloc`` never is called with a second argument less than
+   :macro:`MPZ_MIN_ALLOC` on an *mpz* recieved from :func:`_fmpz_new_mpz`, any
```
Comment (fredrik-johansson) [inline doc/source/fmpz.rst:159]: spelling: received

File: doc/source/fmpz.rst [hunk 2]
```diff
@@ -130,16 +130,35 @@ Types, macros and constants
...
 
-   Clears the ``mpz_t`` "pointed to" by the ``fmpz`` `f`. This is only used
+   Clears the *mpz* "pointed to" by the *fmpz* *f*. This is only used
    internally.
 
+.. note::
+
+   As of FLINT 3.2.0, it is required that the *mpz* pointed to by *f* has at
+   least :macro:`MPZ_MIN_ALLOC` limbs allocated when :func:`_fmpz_clear_mpz` is
+   called. Note that GMP functions never reduces the number of allocated limbs,
```
Comment (fredrik-johansson) [inline doc/source/fmpz.rst:155]: grammar: reduces -> reduce

**Comment by fredrik-johansson** [comment]
Not sure if we *really* want to change all

    mpz_realloc(x, alloc);

to

    if (x->_mp_alloc < alloc)
        mpz_realloc(x, alloc);

as the former can actually free unused memory. This might be desirable, at least in some of the places where it is used.

**Comment by fredrik-johansson** [comment]
> Then that should be handled by the fmpz collector. We have set it to 64 limbs, which I think is a reasonable number.

Yes and no. What can happen in a long running calculation with mixed size numbers is that all cached mpzs eventually saturate the limit and use far more memory than needed when recycled for smaller operands. But if we avoid this problem at all right now, it is probably by accident.

File: doc/source/fmpz.rst (hunk 1 above)
Comment (albinahlback) [reply doc/source/fmpz.rst]: Thanks!

**Comment by albinahlback** [comment]
> > Then that should be handled by the fmpz collector. We have set it to 64 limbs, which I think is a reasonable number.
> 
> Yes and no. What can happen in a long running calculation with mixed size numbers is that all cached mpzs eventually saturate the limit and use far more memory than needed when recycled for smaller operands. But if we avoid this problem at all right now, it is probably by accident.

Sure, I agree with this. But I don't think that is the responsibility of the "smaller" arithmetic routines.

I hope that you can agree with me that the responsibility could fall on
- the memory manager,
- bigger routines (not sure which ones, but say `_fmpz_vec_content_chained`), and
- users who develop important routines which may require harder control of memory management.

**Comment by fredrik-johansson** [comment]
> Sure, I agree with this. But I don't think that is the responsibility of the "smaller" arithmetic routines.

Sure enough, GMP has a ``MPZ_REALLOC`` macro. We could have a similar macro (or inline function). It would then be easy to adjust the global behavior.

**Comment by fredrik-johansson** [comment]
However, consider a case like this in ``fmpz/bit_unpack.c`` (original code):

        /* allocate space for l limbs only */
        mpz_realloc(mcoeff, l);

Here the intent was clearly to resize tightly. Indeed, this makes sense when you are unpacking a big polynomial; in that case, you want to avoid excess allocation for every coefficient.

**Comment by albinahlback** [comment]
Please let me know if you can identify some other functions which may benefit from such reallocations.

**Comment by fredrik-johansson** [comment]
Looks OK except that the names ``MPZ_REALLOC`` and ``MPZ_DO_REALLOC`` don't really suggest what their difference is, and we should be careful about intruding on the GMP namespace. Maybe rename

    MPZ_REALLOC -> FLINT_MPZ_REALLOC or FLINT_MPZ_REALLOC_AT_LEAST
    MPZ_DO_REALLOC -> FLINT_MPZ_REALLOC_TIGHT or FLINT_MPZ_REALLOC

**Comment by albinahlback** [comment]
Good point.

**Comment by albinahlback** [comment]
Will merge if the tests passes.

**Pull request merged**